docs/messages-client-impl.md

```diff
@@ -0,0 +1,92 @@
+
+We then encrypt the payload by JSON-Stringifying the payload then using the Push `aes128gcm` encryption algorithm. See the [Firefox Desktop implementation](https://bugzilla.mozilla.org/show_bug.cgi?id=1442128) or [Web Push Payload Encryption](https://developers.google.com/web/updates/2016/03/web-push-encryption).
+The keys used are the ones of the recipient device. They are obtainable by calling the [GET /accounts/devices](https://github.com/mozilla/fxa-auth-server/blob/master/docs/api.md#get-accountdevices) endpoint (`pushPublicKey`/`pushAuthKey`).
+
+The resulting encrypted durable payload is a *RFC 4648 base64url (no padding)* encoded string.
+
+## (B) Sending the payload
+
+Nothing special here, we use the `sessionToken` authenticated [POST /account/devices/messages](https://github.com/mozilla/fxa-auth-server/blob/master/docs/api.md#post-accountdevicesmessages) endpoint.
+Note that we also specify the `topic` in the request body. The FxA server uses that field to filter messages (*e.g.* Firefox iOS only wants `sendtab`).
+
+## (C) Storing the durable payload
+There's nothing actionable from you on that part. However, you should remember for later that each stored message has an "offset", called `index`.
+
+## (D) Push tickle
+The other device receives a Push notification that looks like this:
+```js
+{
+  "topic": "sendtab",
+  "url": "<URL to retrieve the payload>"
+}
+```
+
+`url` is the direct URL to retrieve the payload, without any consideration for any eventual other pending message.
+
+You have two choices here:
+- Use `url`, just like Firefox iOS does it, in order to guarantee that *1 push message* equals *1 payload retrieved*. However, it is a bit more work to catch out-of-band messages. If you choose that solution you can skip the next section, but pay special attention to the "Periodic polling" section.
+- Do not use `url`, like Firefox Desktop, and treat push notifications as tickles: simply retrieve all pending messages.
+
+## (E) Retrieving the durable payload
+
+We use the `sessionToken` authenticated [GET /account/device/messages](https://github.com/mozilla/fxa-auth-server/blob/master/docs/api.md#get-accountdevicemessages) endpoint.
+An `index` *searchParam* can be set: in that case, the server will return all messages that have been stored after the message who had that `index`.
+You should persist the newly returned index locally, so you don't end-up fetching and replaying the same messages over and over again (messages are NOT deleted after a fetch!).
+
+## (F) Decryption and handling
+
+Decrypt the base64 payload using your own FxA push subscription encryption keys and trigger the appropriate actions!
+
+## Periodic polling
+
+You should call the [GET /account/device/messages](https://github.com/mozilla/fxa-auth-server/blob/master/docs/api.md#get-accountdevicemessages) endpoint periodically in your app (e.g. every 20 minutes/every sync) unconditionally to check for missed messages.
```

Comment:
I remain concerned about the impact this polling will have on the servers, especially from desktop.  Are there any other FxA APIs that desktop calls every sync, so we can take a look at their load impact for comparison?  (I'm thinking e.g. fetching the list of connected devices, but there may be others).

In conversation with @bbangert, he suggested that desktop's push implementation is reliable enough that this periodic polling is not necessary, and the chance of desktop missing a "messageReceived" push notification while the profile is running is basically zero.  I wonder if, at least for desktop, we could do something like:

* At browser startup, attach push message listeners for FxA
* After listeners are attached and any queued push messages have been dispatched, make a single call to `/device/messages` to fetch any that we missed while offline.
* Trust that push will be reliable, and don't poll `/device/messages` while the browser is up and running.